### PR TITLE
feat: custom chart download

### PR DESCRIPTION
**Pull request by Ash-exp**
# Description
**Problem Statement:**
Inability to Download Uploaded Helm Charts on Devtron Hinders Chart Modifications and Re-uploads.

**Solution:**

1. An option to download listed custom charts.
2. Display all devtron default charts in that list.

Fixes https://github.com/devtron-labs/devtron/issues/3840

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] Upload existing Chart and Version
- [x] Upload existing Chart with different version
- [x] Upload new Chart
- [x] Upload Devtron Reference Charts with reserved name
- [x] Upload Devtron Reference Charts with new name
- [x] Upload Chart with editing the description
- [x] Use the Custom Chart in Devtron App

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [x] I have commented my code, particularly in hard-to-understand areas

image
image

### CODE REVIEW

File: src/components/CustomChart/UploadChartModal.tsx [hunk 1]
```diff
@@ -33,7 +33,9 @@ export default function UploadChartModal({ closeUploadPopup }: UploadChartModalT
                 setUploadState(UPLOAD_STATE.ERROR)
                 if (Array.isArray(error.errors)) {
                     if (error.errors[0].code === '5001') {
-                        setErrorData({ title: error.errors[0].userMessage, message: ['Try uploading another chart'] })
+                        setErrorData({ title: error.errors[0]?.userMessage || "", message: ['Try uploading another chart'] })
+                    } else if (error.errors[0].code === '5002') {
```
Comment (shivani170) [inline src/components/CustomChart/UploadChartModal.tsx:37]: 5003 from constant

File: src/components/CustomChart/CustomChartList.tsx [hunk 2]
```diff
@@ -151,21 +135,121 @@ export default function CustomChartList() {
...
+            toast.success('Chart Downloaded Successfully')
+            closeChartVersionsModal()
+        } catch (error) {
+            showError(error)
+        } finally {
+            setDownloadInProgress("")
+        }
+    }
+
+    const randerChartVersionsModalBody = (chartData: ChartDetailType) : JSX.Element => {
```
Comment (shivani170) [inline src/components/CustomChart/CustomChartList.tsx:163]: Typo must be render

File: src/components/CustomChart/CustomChartList.tsx [hunk 3]
```diff
@@ -151,21 +135,121 @@ export default function CustomChartList() {
...
+                <span className="fs-13 fw-6 lh-20">How to use?</span>
+                <span className="fs-13 fw-4 lh-20 ml-4 mr-4">
+                    Uploaded charts can be used in Deployment template to deploy custom applications created in Devtron.
+                </span>
+                {renderLearnMoreLink()}
+            </>
+        )
+    }
+
+    const randerDevtronChipTag = (chartData: ChartDetailType) : JSX.Element => {
```
Comment (shivani170) [inline src/components/CustomChart/CustomChartList.tsx:197]: Typo

File: src/components/CustomChart/customChart.scss [hunk 4]
```diff
@@ -5,7 +5,19 @@
...
+            white-space: nowrap;
+        }
+        .chart-version-row:hover {
+            background-color: var(--N50);
+        }
+    }
+
+    .custom-chart-info-bar {
+        padding: 6px 16px;
+        border-radius: 8px 8px 0px 0px;
```
Comment (shivani170) [inline src/components/CustomChart/customChart.scss:20]: Reduce the CSS

File: src/components/CustomChart/UploadChartModal.tsx (hunk 1 above)
Comment (Ash-exp) [reply src/components/CustomChart/UploadChartModal.tsx]: Fixed 👍

File: src/components/CustomChart/CustomChartList.tsx (hunk 2 above)
Comment (Ash-exp) [reply src/components/CustomChart/CustomChartList.tsx]: Fixed 👍

File: src/components/CustomChart/CustomChartList.tsx (hunk 3 above)
Comment (Ash-exp) [reply src/components/CustomChart/CustomChartList.tsx]: Fixed 👍

File: src/components/CustomChart/customChart.scss (hunk 4 above)
Comment (Ash-exp) [reply src/components/CustomChart/customChart.scss:19]: Fixed 👍

**Review by shivani170 — DISMISSED** [review]

**Review by shivani170 — DISMISSED** [review]

File: yarn.lock
Comment (vivek-devtron) [inline yarn.lock:1]: Revert changes of this file

File: src/components/CustomChart/CustomChartList.tsx [hunk 5]
```diff
@@ -151,21 +135,119 @@ export default function CustomChartList() {
...
+    }
+
+    const handleCustomChartDownload = async(e: any) => {
+        const chartRefId = e.currentTarget.dataset.versionid
+        const chartVersion = e.currentTarget.dataset.version
+        const chartName = e.currentTarget.dataset.name
+        try {
+            setDownloadInProgress(chartName)
+            const a = document.createElement('a')
+            a.href = `${Host}/${Routes.DOWNLOAD_CUSTOM_CHART}/${chartRefId}`
```
Comment (vivek-devtron) [inline src/components/CustomChart/CustomChartList.tsx:149]: We discussed to change this logic. right?

File: yarn.lock
Comment (Ash-exp) [reply yarn.lock:1]: Updated 👍

File: src/components/CustomChart/CustomChartList.tsx (hunk 5 above)
Comment (Ash-exp) [reply src/components/CustomChart/CustomChartList.tsx:122]: Had discussed this with Utkarsh and he suggested that we can keep it this way

**Review by shivani170 — DISMISSED** [review]

**Review by vivek-devtron — DISMISSED** [review]

**Review by vivek-devtron — APPROVED** [review]

**Review by shivani170 — APPROVED** [review]

**Review by shivani170 — DISMISSED** [review]

**Review by vivek-devtron — DISMISSED** [review]

**Review by vivek-devtron — APPROVED** [review]

**Pull request merged**